File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        def twoSum(array, target):
            print("target: ", target)
            left, right = 0, len(array) - 1
            ans = []
            while left < right:
                base = array[left] + array[right]
                if base > target:
                    right -= 1
                elif base < target:
                    left += 1
                else:
                    print(left, right)
                    ans.append([-target, array[left], array[right]])
                    left += 1
                    right -= 1
            return ans

        nums.sort()
        print(nums)
        ans = []
        for i in range(len(nums)):
            if nums[i] > 0:
                return ans
            if i >= 1 and nums[i] == nums[i-1]:
                continue
            ans += twoSum(nums[i+1:], -nums[i])
        return ans

File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        int n = nums.size();
        for (int i = 0; i < n-2; i++) {
            if (i==0 || nums[i] != nums[i-1]) {
                twoSum(nums, i, ans);
            }
        }
        return ans;
    }
private:
    void twoSum(vector<int>& nums, int i, vector<vector<int>>& res) {
        int lo = i+1, hi = nums.size()-1;
        while (lo < hi) {
            int sum = nums[i] + nums[lo] + nums[hi];
            if (sum == 0) {
                res.push_back({nums[i],nums[lo++], nums[hi--]});
                while (lo < hi && nums[lo] == nums[lo-1]) lo++;
            } else if (sum < 0) {
                lo++;
            } else {
                hi--;
            }
        }
    }
};
//class Solution {
//public:
//    vector<vector<int>> threeSum(vector<int>& nums) {
//        vector<vector<int>> ans;
//        sort(nums.begin(), nums.end());
//        int n = nums.size();
//        for (int i = 0; i < n-2; i++) {
//            if (i==0 || nums[i] != nums[i-1]) {
//                twoSum(nums, i, ans);
//            }
//        }
//        return ans; 
//    }
//private:
//    void twoSum(vector<int>& nums, int i, vector<vector<int>>& res) {
//        int lo = i+1, hi = nums.size()-1;
//        while (lo < hi) {
//            int sum = nums[i] + nums[lo] + nums[hi];
//            if (sum < 0) {
//                lo++;
//            } else if (sum > 0) {
//                hi--;
//            } else {
//                res.push_back({nums[i], nums[lo++], nums[hi--]});
//                while (lo < hi && nums[lo] == nums[lo-1]) lo++;
//            }
//        }
//    }
//};